STM32/DC/Core/Src/main.c

```diff
@@ -151,9 +154,25 @@ void printHeader()
 }
 
 
+void shutOffAndReport(bool isCurrentExceeded){
+	allOff();
+	if (isCurrentExceeded){
+		USBnprintf("warning: max current draw exceeded.");
+	} else {
+		USBnprintf("warning: max temp exceeded.");
+	}
+}
+
 static double meanCurrent(const int16_t *pData, uint16_t channel)
 {
-    return current_scalar * ADCMean(pData, channel) + current_bias;
+    double meanCurrent = current_scalar * ADCMean(pData, channel) + current_bias;
+
+    // If component draws more current than max allowed then operation is shut off
+    if (meanCurrent > MAX_CURRENT_DRAW){
+    	bool isCurrentExceeded = true;

```

Comment:
I am not a big fan of putting this shut off logic in the simple functions meanCurrent etc.
When looking at the name of function it's indicate that it performs a simple calculation. Now it is also handling system logic.

Suggest to save values in printValues. This will also make it possible to list which valve it drawing to much current/temp.